waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+       :items
+       (mapv replicaset->Service)))
+
+(defn- get-replicaset-pods
+  [api-server-url http-client {:keys [app-name namespace] :as service}]
+  (->> (str api-server-url
+            "/api/v1/namespaces/"
+            namespace
+            "/pods?labelSelector=app="
+            app-name)
+       (api-request http-client)
+       :items))
+
+(defn- get-service-instances
+  [api-server-url http-client service]
+  (vec (for [pod (get-replicaset-pods api-server-url http-client service)
+             :let [service-instance (pod->ServiceInstance pod)]
+             :when (:host service-instance)]
+         (doto service-instance
+           (track-failed-instances pod)))))
+
+(defn- instances-breakdown
+  [api-server-url http-client service]
+  {:active-instances (get-service-instances api-server-url http-client service)
+   :failed-instances (get @service-id->failed-instances-transient-store (:id service) [])})
+
+(defn- service+instances
+  [api-server-url http-client service]
+  (let [instances (instances-breakdown api-server-url http-client service)]
+    [service instances]))
+
+(defn- patch-object-json
+  [k8s-object-uri http-client ops]
+  (api-request http-client k8s-object-uri
+               :request-method :patch
+               :content-type "application/json-patch+json"
+               :body (as-json ops)))
+
+(defn- patch-object-replicas
+  [k8s-object-uri http-client replicas replicas']
+  (patch-object-json http-client k8s-object-uri
+                     ;; NOTE: ~1 is JSON-patch escape syntax for a "/" in a key name
+                     ;; see http://jsonpatch.com/#json-pointer
+                     [{:op :test, :path "/metadata/annotations/waiter~1app-status", :value "live"}
+                      {:op :test, :path "/spec/replicas", :value replicas}
+                      {:op :replace, :path "/spec/replicas", :value replicas'}]))
+
+(defn- scale-service-to
+  [api-server-url http-client service instances']
+  (log/info "Scaling to " instances' service)
+  (let [request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (:namespace service)
+                         "/replicasets/"
+                         (:app-name service))
+        instances (:task-count service)]
+    (patch-object-replicas http-client request-url instances instances')))
+
+(defn- scale-service-by
+  [api-server-url http-client service instances-delta]
+  (log/info "Scaling incrementally by " instances-delta service)
+  (let [request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (:namespace service)
+                         "/replicasets/"
+                         (:app-name service))
+        instances (:task-count service)
+        instances' (+ instances instances-delta)]
+    (patch-object-replicas http-client request-url instances instances')))
+
+(defn- kill-service-instance
```

Comment:
Does there need to be a wait before requesting termination and force killing?

Reply:
Do we need additional logic to handle the scenario if there is an exception thrown between any of the term, scale-down, kill statements?

Reply:
> Does there need to be a wait before requesting termination and force killing?

No, that's not needed. I'm writing up some documentation explaining this.

> Do we need additional logic to handle the scenario if there is an exception thrown between any of the term, scale-down, kill statements?

Yes, that's one of my TODO items listed at the top of this PR.